src/btree/bt_discard.c

```diff
@@ -245,20 +245,31 @@ __free_page_modify(WT_SESSION_IMPL *session, WT_PAGE *page)
 void
 __wt_ref_addr_free(WT_SESSION_IMPL *session, WT_REF *ref)
 {
+    WT_PAGE *home;
     void *ref_addr;
 
+    home = ref->home;
+
     /*
-     * The page being discarded may be the child of a page being split, where the WT_REF.addr field
-     * is being instantiated (as it can no longer reference the on-disk image). Loop until we read
-     * and clear the address without a race, then free the read address as necessary.
+     * In order to free the WT_REF.addr field we need to read and clear the address without a race.
+     * The WT_REF may be a child of a page being split, in which case the addr field could be
+     * instantiated concurrently which changes the addr field. Once we swap in NULL we effectively
+     * own the addr. Then provided the addr is off page we can free the memory.
+     *
+     * However as we could be the child of a page being split the ref->home pointer which tells us
+     * whether the addr is on or off page could change concurrently. To avoid this we save the home
+     * pointer before we do the compare and swap.
```

Comment:
I think it'd be worth including that `ref->home` being volatile stops the compiler rewriting this with multiple reads (the original code was slightly wrong in the regard).

Reply:
[a441a1a](https://github.com/wiredtiger/wiredtiger/pull/8870/commits/a441a1a6c4e78a4a86298c0474520dc6f75a8ba5)

Reply:
Thanks, one last issue from me - that commit should have `compiler` rather than `compile`.

Reply:
f9981be4a